F5 Bigip Cipher Rules and Cipher Groups

There is no way I can create ltm cipher groups nor  cipher rules , is this something i can expect  in a week or two .

@RavinderReddyF5 @pgouband 
Testing the feature with v1.20.0 and the results are not good.

I have the following code:
```
resource "bigip_ltm_cipher_rule" "test_cipher_rule" {
  name                 = "/Common/test_cipher_rule"
  cipher               = "TLS13-AES128-GCM-SHA256:TLS13-AES256-GCM-SHA384"
  dh_groups            = "P256:P384:FFDHE2048:FFDHE3072:FFDHE4096"
  signature_algorithms = "DEFAULT"
}

resource "bigip_ltm_cipher_group" "test-cipher-group" {
  name = "/Common/test-cipher-group-01"
  allow = ["/Common/f5-aes"]
  require = ["/Common/f5-quic"]
  ordering = "speed"
}
```

Plan result looks good:
```
Terraform will perform the following actions:

  # bigip_ltm_cipher_group.test-cipher-group will be created
  + resource "bigip_ltm_cipher_group" "test-cipher-group" {
      + allow    = [
          + "/Common/f5-aes",
        ]
      + id       = (known after apply)
      + name     = "/Common/test-cipher-group-01"
      + ordering = "speed"
      + require  = [
          + "/Common/f5-quic",
        ]
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```
Apply went well and on the UI, the config reflects my TF code.

Then I change my code for the cipher_group to:
```
resource "bigip_ltm_cipher_group" "test-cipher-group" {
  name = "/Common/test-cipher-group-01"
  allow = [resource.bigip_ltm_cipher_rule.test_cipher_rule.name]
}
```

Plan result (**not OK** because the ordering should be set back to its default value="DEFAULT"):
```
Terraform will perform the following actions:

  # bigip_ltm_cipher_group.test-cipher-group will be updated in-place
  ~ resource "bigip_ltm_cipher_group" "test-cipher-group" {
      ~ allow    = [
          - "/Common/f5-aes",
          + "/Common/test_cipher_rule",
        ]
        id       = "/Common/test-cipher-group-01"
        name     = "/Common/test-cipher-group-01"
      ~ require  = [
          - "/Common/f5-quic",
        ]
        # (1 unchanged attribute hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
Result after apply on the UI is not OK:
1. the allowed cipher has been successfully changed from f5-aes to test_cipher_rule but the require (restrict) still has the f5-quic configured
2. the ordering is still "Speed" and not "DEFAULT" (as my previous remark)

If I run te plan again, it detects that the f5-quic has to be removed:

```
Terraform will perform the following actions:

  # bigip_ltm_cipher_group.test-cipher-group will be updated in-place
  ~ resource "bigip_ltm_cipher_group" "test-cipher-group" {
        id       = "/Common/test-cipher-group-01"
        name     = "/Common/test-cipher-group-01"
      ~ require  = [
          - "/Common/f5-quic",
        ]
        # (2 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Another issue: if I remove (through the UI) that cipher_group and run a plan, I crash the provider:
```
Stack trace from the terraform-provider-bigip_v1.20.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9f42e0]

goroutine 796 [running]:
github.com/F5Networks/terraform-provider-bigip/bigip.resourceBigipLtmCipherGroupRead({0xfa6960?, 0xc00008b7a0?}, 0xc000da0500, {0xe2a8c0?, 0xc00050ba80})
        github.com/F5Networks/terraform-provider-bigip/bigip/resource_bigip_ltm_cipher_group.go:114 +0x180
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0004490a0, {0xfa6960, 0xc00008b7a0}, 0xd?, {0xe2a8c0, 0xc00050ba80})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:724 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0004490a0, {0xfa6960, 0xc00008b7a0}, 0xc000688c30, {0xe2a8c0, 0xc00050ba80})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:1015 +0x585
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc0003a4c00, {0xfa6960?, 0xc00008b680?}, 0xc0006f4880)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/grpc_provider.go:613 +0x4a5
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc000000500, {0xfa6960?, 0xc00008aed0?}, 0xc00067e600)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:748 +0x4b1
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xdecf80?, 0xc000000500}, {0xfa6960, 0xc00008aed0}, 0xc000772a80, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:349 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003bc000, {0xfaa500, 0xc0002ce820}, 0xc00014b7a0, 0xc000444840, 0x1513e10, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1336 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0003bc000, {0xfaa500, 0xc0002ce820}, 0xc00014b7a0, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1704 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-bigip_v1.20.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

```

--- a/bigip/resource_bigip_ltm_cipher_group.go
+++ b/bigip/resource_bigip_ltm_cipher_group.go
@@ -41,23 +41,20 @@ func resourceBigipLtmCipherGroup() *schema.Resource {
 				Description: "Specifies descriptive text that identifies the cipher rule",
 			},
 			"ordering": {
-				Type:     schema.TypeString,
-				Optional: true,
-				Computed: true,
-				//Default:     "default",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "default",
 				Description: "Controls the order of the Cipher String list in the Cipher Audit section. Options are Default, Speed, Strength, FIPS, and Hardware. The rules are processed in the order listed",
 			},
 			"allow": {
 				Type:        schema.TypeSet,
 				Optional:    true,
-				Computed:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Specifies the configuration of the allowed groups of ciphers. You can select a cipher rule from the Available Cipher Rules list",
 			},
 			"require": {
 				Type:        schema.TypeSet,
 				Optional:    true,
-				Computed:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Specifies the configuration of the restrict groups of ciphers. You can select a cipher rule from the Available Cipher Rules list",
 			},
@@ -155,11 +152,6 @@ func resourceBigipLtmCipherGroupUpdate(ctx context.Context, d *schema.ResourceDa
 	if err := client.ModifyLtmCipherGroupNew(name, new); err != nil {
 		return diag.FromErr(fmt.Errorf("error modifying cipher group %s: %v", name, err))
 	}
-
-	//
-	// if err := client.ModifyLtmCipherGroup(name, cipherGroupconfig); err != nil {
-	//	return diag.FromErr(fmt.Errorf("error modifying cipher group %s: %v", name, err))
-	// }
 
 	return resourceBigipLtmCipherGroupRead(ctx, d, meta)
 }

--- a/bigip/resource_bigip_ltm_cipher_group_test.go
+++ b/bigip/resource_bigip_ltm_cipher_group_test.go
@@ -43,6 +43,54 @@ func TestAccBigipLtmCipherGroupCreateTC1(t *testing.T) {
 	})
 }
 
+func TestAccBigipLtmCipherGroupRemoveRequire(t *testing.T) {
+	cipherGrpCfg := `
+resource "bigip_ltm_cipher_group" "test-cipher-group" {
+  name     = "/Common/testciphergrp"
+  allow    = ["/Common/f5-aes"]
+  %s
+}
+`
+	requireAndOrdering := `
+  require  = ["/Common/f5-quic"]
+  ordering = "speed"
+`
+
+	c1 := fmt.Sprintf(cipherGrpCfg, requireAndOrdering)
+	c2 := fmt.Sprintf(cipherGrpCfg, "")
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: c1,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckCipherGroupExists("/Common/testciphergrp"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "name", "/Common/testciphergrp"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "allow.#", "1"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "require.#", "1"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "allow.0", "/Common/f5-aes"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "require.0", "/Common/f5-quic"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "ordering", "speed"),
+				),
+			},
+			{
+				Config: c2,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckCipherGroupExists("/Common/testciphergrp"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "name", "/Common/testciphergrp"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "allow.#", "1"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "require.#", "0"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "allow.0", "/Common/f5-aes"),
+					resource.TestCheckResourceAttr("bigip_ltm_cipher_group.test-cipher-group", "ordering", "default"),
+				),
+			},
+		},
+	})
+}
+
 func testCheckCipherGroupExists(name string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		client := testAccProvider.Meta().(*bigip.BigIP)
@@ -52,7 +100,7 @@ func testCheckCipherGroupExists(name string) resource.TestCheckFunc {
 			return err
 		}
 		if p == nil {
-			return fmt.Errorf("Pool %s does not exist ", name)
+			return fmt.Errorf("cipher group %s does not exist ", name)
 		}
 
 		return nil
@@ -66,11 +114,11 @@ func testCheckCipherGroupDestroyed(s *terraform.State) error {
 			continue
 		}
 		name := rs.Primary.ID
-		pool, err := client.GetLtmCipherGroup(name)
+		cipherGroup, err := client.GetLtmCipherGroup(name)
 		if err != nil {
 			return err
 		}
-		if pool != nil {
+		if cipherGroup != nil {
 			return fmt.Errorf("Cipher rule %s not destroyed ", name)
 		}
 	}

--- a/docs/resources/bigip_ltm_cipher_group.md
+++ b/docs/resources/bigip_ltm_cipher_group.md
@@ -3,11 +3,10 @@ layout: "bigip"
 page_title: "BIG-IP: bigip_ltm_cipher_group"
 subcategory: "Local Traffic Manager(LTM)"
 description: |-
-Provides details about bigip_ltm_cipher_group resource
+  Provides details about bigip_ltm_cipher_group resource
 ---
 
 # bigip\_ltm\_cipher\_group
-
 `bigip_ltm_cipher_group` Manages F5 BIG-IP LTM cipher group using iControl REST.
 
 ## Example Usage
@@ -25,11 +24,11 @@ resource "bigip_ltm_cipher_group" "test-cipher-group" {
 
 * `name` - (Required,type `string`) Name of the Cipher group. Name should be in pattern `partition` + `cipher_group_name`
 
-* `allow` - (Optional,type `list` of `strings` ) Specifies the configuration of the allowed groups of ciphers. You can select a cipher rule from the Available Cipher Rules list.
+* `allow` - (Optional,type `list` of `strings` ) Specifies the configuration of the allowed groups of ciphers. You can select a cipher rule from the Available Cipher Rules list. To have no allowed ciphers, omit this attribute in the config or set it to an empty set like, `[]`.
 
-* `require` - (Optional,type `list` of `string`) Specifies the configuration of the restrict groups of ciphers. You can select a cipher rule from the Available Cipher Rules list.
+* `require` - (Optional,type `list` of `string`) Specifies the configuration of the restrict groups of ciphers. You can select a cipher rule from the Available Cipher Rules list. To have no restricted ciphers, omit this attribute in the config or set it to an empty set like, `[]`.
 
-* `ordering` - (Optional,type `string`) Controls the order of the Cipher String list in the Cipher Audit section. Options are Default, Speed, Strength, FIPS, and Hardware. The rules are processed in the order listed.
+* `ordering` - (Optional,type `string`) Controls the order of the Cipher String list in the Cipher Audit section. Options are Default, Speed, Strength, FIPS, and Hardware. The rules are processed in the order listed. The default is `default`.
 
 ## Importing
 An existing cipher group can be imported into this resource by supplying the cipher rule full path name ex : `/partition/name`

